service/history/api/updateworkflow/api.go

```diff
@@ -125,7 +125,7 @@ func Invoke(
 			if upd, alreadyExisted, err = updateReg.FindOrCreate(ctx, updateID); err != nil {
 				return nil, err
 			}
-			if err = upd.OnMessage(ctx, req.GetRequest().GetRequest(), workflow.WithEffects(effect.Immediate(ctx), ms)); err != nil {
+			if err = upd.OnMessage(ctx, req.GetRequest().GetRequest(), ms.IsWorkflowExecutionRunning(), workflow.WithEffects(effect.Immediate(ctx), ms)); err != nil {
```

Comment:
If when the update request is received the workflow is already closed, we will return NotFound error (on L96 above), but if workflow is closed after update is received but before it's sent, we will reject the update with some message. I think it will be ideal to make them consistent.

Reply:
yeah that SGTM.